rosbag2_transport/test/rosbag2_transport/test_play.cpp

```diff
@@ -119,3 +120,49 @@ TEST_F(RosBag2PlayTestFixture, recorded_messages_are_played_for_all_topics)
           &test_msgs::msg::Arrays::float32_values,
           ElementsAre(40.0f, 2.0f, 0.0f)))));
 }
+
+TEST_F(RosBag2PlayTestFixture, topic_qos_profiles_overriden)
+{
+  const auto topic_name = "/test_topic";
+  const auto msg_type = "test_msgs/BasicTypes";
+  const auto num_msgs = 3;
+  auto basic_msg = get_messages_basic_types()[0];
+  basic_msg->int32_value = 42;
+  const auto topic_types = std::vector<rosbag2_storage::TopicMetadata>{
+    {topic_name, msg_type, "" /*serialization_format*/, "" /*offered_qos_profiles*/}
+  };
+
+  const std::vector<int64_t> topic_timestamps_ms = {500, 700, 900};
+  std::vector<std::shared_ptr<rosbag2_storage::SerializedBagMessage>> messages;
+  messages.reserve(topic_timestamps_ms.size());
+  for (const auto topic_timestamp : topic_timestamps_ms) {
+    messages.push_back(serialize_test_message(topic_name, topic_timestamp, basic_msg));
+  }
+
+  auto prepared_mock_reader = std::make_unique<MockSequentialReader>();
+  prepared_mock_reader->prepare(messages, topic_types);
+  reader_ = std::make_unique<rosbag2_cpp::Reader>(std::move(prepared_mock_reader));
+
+  const auto qos_request = rclcpp::QoS{rclcpp::KeepAll()}.durability_volatile();
+  sub_->add_subscription<test_msgs::msg::BasicTypes>(topic_name, num_msgs, qos_request);
+  auto await_received_messages = sub_->spin_subscriptions();
+
+  // The previous subscriber requested durability VOLATILE which is the default in rosbag2.
+  // We override the requested durability to TRANSIENT_LOCAL so that we can receive messages.
+  // If the previous subscription requested TRANSIENT_LOCAL and we overrode with VOLATILE, then we
+  // would not receive any messages.
```

Comment:
We can add "wait for condition" checks with timeouts to check that "something doesn't happen". My only complaint with these tests is that they take a minimum of.. say 3 seconds as a reasonable timeout, so the test suite starts to take a long time when we have a lot of them. I can't think of a better way, though, given the mechanics of DDS discovery...